clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java

```diff
@@ -0,0 +1,364 @@
+            logContext,
+            time,
+            0,
+            RETRY_BACKOFF_MS,
+            RETRY_BACKOFF_MAX_MS,
+            0);
+        heartbeatRequestManager = createManager();
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
+        when(membershipManager.shouldSendHeartbeat()).thenReturn(true);
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(time.milliseconds());
+        assertEquals(1, result.unsentRequests.size());
+        result.unsentRequests.get(0).future().completeExceptionally(new TimeoutException("timeout"));
+
+        // assure the manager will backoff on timeout
+        time.sleep(RETRY_BACKOFF_MS - 1);
+        result = heartbeatRequestManager.poll(time.milliseconds());
+        assertEquals(0, result.unsentRequests.size());
+
+        time.sleep(1);
+        result = heartbeatRequestManager.poll(time.milliseconds());
+        assertEquals(1, result.unsentRequests.size());
+    }
+
+    @Test
+    public void testNoCoordinator() {
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(time.milliseconds());
+
+        assertEquals(Long.MAX_VALUE, result.timeUntilNextPollMs);
+        assertEquals(0, result.unsentRequests.size());
+    }
+
+    @ParameterizedTest
+    @ApiKeyVersionsSource(apiKey = ApiKeys.CONSUMER_GROUP_HEARTBEAT)
+    public void testValidateConsumerGroupHeartbeatRequest(final short version) {
+        heartbeatRequestState = new HeartbeatRequestManager.HeartbeatRequestState(
+            logContext,
+            time,
+            0,
+            RETRY_BACKOFF_MS,
+            RETRY_BACKOFF_MAX_MS,
+            0);
+        heartbeatRequestManager = createManager();
+        // Update membershipManager's memberId and memberEpoch
+        ConsumerGroupHeartbeatResponse result =
+            new ConsumerGroupHeartbeatResponse(new ConsumerGroupHeartbeatResponseData()
+            .setMemberId(memberId)
+            .setMemberEpoch(memberEpoch));
+        membershipManager.updateState(result.data());
+
+        // Create a ConsumerHeartbeatRequest and verify the payload
+        NetworkClientDelegate.PollResult pollResult = heartbeatRequestManager.poll(time.milliseconds());
+        assertEquals(1, pollResult.unsentRequests.size());
+        NetworkClientDelegate.UnsentRequest request = pollResult.unsentRequests.get(0);
+        assertTrue(request.requestBuilder() instanceof ConsumerGroupHeartbeatRequest.Builder);
+
+        ConsumerGroupHeartbeatRequest heartbeatRequest =
+            (ConsumerGroupHeartbeatRequest) request.requestBuilder().build(version);
+        assertEquals(GROUP_ID, heartbeatRequest.data().groupId());
+        assertEquals(memberId, heartbeatRequest.data().memberId());
+        assertEquals(memberEpoch, heartbeatRequest.data().memberEpoch());
```

Comment:
Should we verify all the fields? We also set others [here](https://github.com/apache/kafka/pull/14364/files#diff-0eb94945bb75c3b1bda527708a48af59eafafc5bd9f85136fff0d0c405590497R166).

Reply:
Done - I left a todo to verify client and server side assignors and pattern regex.